src/MmappedFileMonitor.cc

```diff
@@ -0,0 +1,67 @@
+/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include "MmappedFileMonitor.h"
+
+#include "RecordSession.h"
+#include "RecordTask.h"
+#include "log.h"
+
+using namespace std;
+
+namespace rr {
+
+MmappedFileMonitor::MmappedFileMonitor(Task* t, int fd) {
+  ASSERT(t, !t->session().is_replaying());
+  stat_ = t->stat_fd(fd);
+}
+
+MmappedFileMonitor::MmappedFileMonitor(Task* t, EmuFile::shr_ptr f) {
+  ASSERT(t, t->session().is_replaying());
+  emu_file_ = f;
+}
+
+void MmappedFileMonitor::did_write(Task* t, const std::vector<Range>& ranges,
+				   int64_t offset) {
+  if (ranges.empty()) {
+    return;
+  }
+
+  ASSERT(t, offset >= 0)
+      << "Only pwrite/pwritev supported on /proc/<pid>/mem currently";
+  // XXX to fix that, we'd have to grab the file offset from
+  // /proc/<pid>/fdinfo.
+
+  bool is_replay = t->session().is_replaying();
+  for (auto m : t->vm()->maps()) {
```

Comment:
Technically we should scan the mappings in other processes also. That doesn't seem too hard to do? Just loop over all tasks and keep track of which AddressSpaces you've already seen so we don't process them multiple times.